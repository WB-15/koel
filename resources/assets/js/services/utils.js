export default {
    /**
     * Convert a duration in seconds into H:i:s format.
     * If H is 0, it will be ommited.
     */
    secondsToHis(d) {
        d = parseInt(d);
        
        var s = d%60;

        if (s < 10) {
            s = '0' + s;
        }

        var i = Math.floor((d/60)%60);

        if (i < 10) {
            i = '0' + i;
        }

        var h = Math.floor(d/3600);

        if (h < 10) {
            h = '0' + h;
        }

        return (h === '00' ? '' : h + ':') + i + ':' + s;
    },

    /**
     * Parse the validation error from the server into a flattened array of messages.
     * 
     * @param  {Object}  error  The error object in JSON format.
     *                                    
     * @return {Array}
     */
    parseValidationError(error) {
        return Object.keys(error).reduce((messages, field) => messages.concat(error[field]), []);
    }
};
