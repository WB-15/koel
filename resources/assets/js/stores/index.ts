export * from './albumStore'
export * from './artistStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './playlistStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './settingStore'
export * from './commonStore'
export * from './songStore'
export * from './userStore'
export * from './searchStore'
export * from './themeStore'
