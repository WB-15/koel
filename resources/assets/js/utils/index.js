export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
