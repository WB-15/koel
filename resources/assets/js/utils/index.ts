export * from './event'
export * from './alerts'
export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
export * from './$'
export * from './helpers'
export * from './fileReader'
export * from './directoryReader'
