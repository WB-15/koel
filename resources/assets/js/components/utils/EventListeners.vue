<template>
  <span></span>
</template>

<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { authService } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.is_smart && !playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    authService.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
