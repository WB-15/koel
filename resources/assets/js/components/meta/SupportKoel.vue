<template>
  <div class="support-bar" v-if="shown">
    <p>
      Loving Koel? Please consider supporting its development via
      <a href="https://github.com/users/phanan/sponsorship" rel="noopener" target="_blank">GitHub Sponsors</a>
      and/or
      <a href="https://opencollective.com/koel" rel="noopener" target="_blank">OpenCollective</a>.
    </p>
    <button @click.prevent="close">Hide</button>
    <span class="sep"></span>
    <button @click.prevent="stopBugging">Don't bug me again</button>
  </div>
</template>

<script lang="ts" setup>
import isMobile from 'ismobilejs'
import { computed, ref } from 'vue'
import { eventBus } from '@/utils'
import { preferenceStore as preferences } from '@/stores'

const DELAY_UNTIL_SHOWN = 30 * 60 * 1000
let SUPPORT_BAR_TIMEOUT_HANDLE = 0

const shown = ref(false)

const canNag = computed(() => !isMobile.any && !preferences.supportBarNoBugging)

const setUpShowBarTimeout = () => {
  SUPPORT_BAR_TIMEOUT_HANDLE = window.setTimeout(() => (shown.value = true), DELAY_UNTIL_SHOWN)
}

const close = () => {
  shown.value = false
  window.clearTimeout(SUPPORT_BAR_TIMEOUT_HANDLE)
}

const stopBugging = () => {
  preferences.supportBarNoBugging = true
  close()
}

eventBus.on('KOEL_READY', () => canNag.value && setUpShowBarTimeout())
</script>

<style lang="scss" scoped>
.support-bar {
  background: var(--color-bg-primary);
  font-size: .9rem;
  padding: .75rem 1rem;
  display: flex;
  color: rgba(255, 255, 255, .6);
  z-index: 9;

  > * + * {
    margin-left: 1rem;
  }

  p {
    flex: 1;
  }

  a {
    color: var(--color-text-primary);

    &:hover {
      color: var(--color-highlight);
    }
  }

  .sep {
    display: block;

    &::after {
      content: 'â€¢';
      display: block;
    }
  }

  button {
    color: var(--color-text-primary);
    font-size: .9rem;

    &:hover {
      color: var(--color-highlight);
    }
  }
}
</style>
