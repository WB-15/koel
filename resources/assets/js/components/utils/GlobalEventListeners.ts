/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */

import { defineComponent } from 'vue'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow, requireInjection } from '@/utils'
import { RouterKey } from '@/symbols'
import { useDialogBox, useMessageToaster } from '@/composables'

export const GlobalEventListeners = defineComponent({
  setup (props, { slots }) {
    const { toastSuccess } = useMessageToaster()
    const { showConfirmDialog } = useDialogBox()
    const router = requireInjection(RouterKey)

    eventBus.on('PLAYLIST_DELETE', async playlist => {
      if (await showConfirmDialog(`Delete the playlist "${playlist.name}"?`)) {
        await playlistStore.delete(playlist)
        toastSuccess(`Playlist "${playlist.name}" deleted.`)
        router.go('home')
      }
    }).on('PLAYLIST_FOLDER_DELETE', async folder => {
      if (await showConfirmDialog(`Delete the playlist folder "${folder.name}"?`)) {
        await playlistFolderStore.delete(folder)
        toastSuccess(`Playlist folder "${folder.name}" deleted.`)
        router.go('home')
      }
    }).on('LOG_OUT', async () => {
      await userStore.logout()
      authService.destroy()
      forceReloadWindow()
    })

    return () => slots.default?.()
  }
})
