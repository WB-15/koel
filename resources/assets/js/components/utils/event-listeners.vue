<template>
  <span></span>
</template>
<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { auth } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    auth.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
