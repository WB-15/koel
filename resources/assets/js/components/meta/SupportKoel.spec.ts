import { afterEach, beforeEach, expect, it, vi } from 'vitest'
import { cleanup, fireEvent } from '@testing-library/vue'
import { render } from '@/__tests__/__helpers__'
import { eventBus } from '@/utils'
import { nextTick } from 'vue'
import { preferenceStore } from '@/stores'
import SupportKoel from './SupportKoel.vue'

beforeEach(() => {
  cleanup()
  vi.useFakeTimers()
})

afterEach(() => {
  vi.useRealTimers()
  preferenceStore.state.supportBarNoBugging = false
})

const mountComponent = async () => {
  const result = render(SupportKoel)
  eventBus.emit('KOEL_READY')

  vi.advanceTimersByTime(30 * 60 * 1000)
  await nextTick()

  return result
}

it('shows after a delay', async () => {
  const { html } = await mountComponent()

  expect(html()).toMatchSnapshot()
})

it('does not show if user so demands', async () => {
  preferenceStore.state.supportBarNoBugging = true
  const { queryByTestId } = await mountComponent()

  expect(await queryByTestId('support-bar')).toBe(null)
})

it('hides', async () => {
  const { getByTestId, queryByTestId } = await mountComponent()

  await fireEvent.click(getByTestId('hide-support-koel'))

  expect(await queryByTestId('support-bar')).toBe(null)
})

it('hides and does not bug again', async () => {
  const { getByTestId, queryByTestId } = await mountComponent()

  await fireEvent.click(getByTestId('stop-support-koel-bugging'))

  expect(await queryByTestId('btn-stop-support-koel-bugging')).toBe(null)
  expect(preferenceStore.state.supportBarNoBugging).toBe(true)
})
