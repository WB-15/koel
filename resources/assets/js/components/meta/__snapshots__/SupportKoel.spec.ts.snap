// Vitest Snapshot v1

exports[`shows after a delay 1`] = `
"<div class=\\"support-bar\\" data-testid=\\"support-bar\\" data-v-0db4e332=\\"\\">
  <p data-v-0db4e332=\\"\\"> Loving Koel? Please consider supporting its development via <a href=\\"https://github.com/users/phanan/sponsorship\\" rel=\\"noopener\\" target=\\"_blank\\" data-v-0db4e332=\\"\\">GitHub Sponsors</a> and/or <a href=\\"https://opencollective.com/koel\\" rel=\\"noopener\\" target=\\"_blank\\" data-v-0db4e332=\\"\\">OpenCollective</a>. </p><button data-testid=\\"hide-support-koel\\" data-v-0db4e332=\\"\\">Hide</button><span class=\\"sep\\" data-v-0db4e332=\\"\\"></span><button data-testid=\\"stop-support-koel-bugging\\" data-v-0db4e332=\\"\\">Don't bug me again</button>
</div>"
`;
